docs/proposals/003-applications-outside-argocd-namespace.md

```diff
@@ -0,0 +1,418 @@
+this process to be in full control of my DevOps team.
+
+As a cluster admin, I want to allow my users the self management of Argo CD
+`Application` resources without getting involved in the creation process,
+e.g. by reviewing & approving PRs into our admin namespace. I still want
+to be sure that users cannot circumvent any restrictions that our organisation
+wants to impose on these applications capabilities.
+
+#### Use case 2: App-of-apps pattern for my own applications
+
+As a developer, I want to have the ability to use app-of-apps for my own
+`Application` resources. For this, I will have to create `Application`
+manifests, but I'm currently not allowed to write to the `argocd` namespace.
+
+#### Use case 3: Easy onboarding of new applications and tenants
+
+As an administrator, I want to provide my tenants with a very easy way to
+create their applications from a simple commit to Git, without losing my
+ability to govern and restrict what goes into my cluster.  I want to set up
+an Argo CD application that reconciles my tenant's `Application` manifests to
+a fixed location (namespace) in my cluster, so that the tenant can just put
+their manifests into the Git repository and Argo CD will pick it up from
+there, without having to use complex tools such as Open Policy Agent to
+enforce what AppProjects my tenants are allowed to use.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Application names
+
+One major challenge to solve is the _uniqueness_ of `Application` names. As of
+now this is enforced by Kubernetes, since the `Application` name is also the
+name of the Kubernetes resource (e.g. value of `.metadata.name`). Because all
+`Application` resources must currently exist in the same namespace, Kubernetes
+already enforces a unique name for each `Application`.
+
+This imposes a major challenge when `Application` resources may exist in other
+namespaces, because it would be possible that two resources will have the same
+name.
+
+One way to mitigate this would be to decouple the application's name from the
+name of the `Application` resource by considering the name of the resource's
+namespace as part of the applicatio name.
+
+E.g. consider the two following examples:
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: some-app
+  namespace: foons
+spec:
+  project: some-project
+```
+
+and
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: some-app
+  namespace: barns
+spec:
+  project: some-project
+```
+
+The _name_ of the first application would become `foons/some-app` and the _name_
+of the second application would become `barns/some-app`. This would have the
+advantage to instantly have a clue about _where_ the application is coming from,
+without having to resort to the Kubernetes API to inspect `.spec.namespace`
+field of the app. The notation may also use a hyphen (`-`) or underscore (`_`)
```

Comment:
a `-` should not be used as it is allowed in k8s names. See the two following conflicting examples:

```
name: hello
namespace: hi-all
---
name: hello-hi
namespace: all
```

both will be `hello-hi-all` if separator be a dash.

Reply:
Good catch. Thinking about this for a while :) Will come back to it later.